src/Controls/src/Core/Page/Page.cs

```diff
@@ -922,5 +922,123 @@ private protected override string GetDebuggerDisplay()
 			var debugText = DebuggerDisplayHelpers.GetDebugText(nameof(BindingContext), BindingContext, nameof(Title), Title);
 			return $"{this.GetType().FullName}: {debugText}";
 		}
+
+#nullable enable
+		IDisposable? _handleNavigatedEventsForRootPage = null!;
+
+		internal void WireUpAsOutgoingPage(Page? outgoingPage, NavigationType navigationType)
+		{
+			_handleNavigatedEventsForRootPage?.Dispose();
+			_handleNavigatedEventsForRootPage = null;
+
+			if (HasNavigatedTo)
+			{
+				SendNavigatingFrom(new NavigatingFromEventArgs());
+			}
+
+			if (IsLoaded)
+			{
+				_handleNavigatedEventsForRootPage =
+					this.OnUnloaded(() =>
+					{
+						if (HasNavigatedTo)
+						{
+							SendNavigatedFrom(new NavigatedFromEventArgs(outgoingPage, navigationType));
+						}
+
+						outgoingPage?.DisconnectHandlers();
+						_handleNavigatedEventsForRootPage?.Dispose();
+						_handleNavigatedEventsForRootPage = null;
+						outgoingPage = null;
+					});
+			}
+			else
+			{
+				if (HasNavigatedTo)
+				{
+					SendNavigatedFrom(new NavigatedFromEventArgs(outgoingPage, navigationType));
+				}
+
+				outgoingPage?.DisconnectHandlers();
+			}
+		}
+
+		internal void WireUpAsIncomingPage(Page? oldPage)
+		{
+			oldPage?.WireUpAsOutgoingPage(oldPage, NavigationType.PageSwap);
+			_handleNavigatedEventsForRootPage?.Dispose();
+			_handleNavigatedEventsForRootPage = null;
+
+			IDisposable? newPageUnloaded = null;
+			IDisposable? newPageLoaded = null;
+
+			var previousPage = oldPage;
```

Comment:
When this is called again as part of the "rewire up the events to watch for unloaded" pass, this will be null. Is that OK?

Is there a better way we can name these methods or something? Maybe it is just me, but the double calls and similar code but supposed different behavior is whacking my brains. Maybe I am tired, so will re-review tomorrow just in case.

Reply:
No, it confused me too when I wrote it :-p 

That was one reason I hadn't come back to it, I wasn't happy with the naming.

Also, I think the implementation of `WireUpAsOutgoingPage` is wrong (as I noted above)